0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,159 @@
+
+## Flow
+
+Before the protocol starts, the sender and receiver use a higher-level protocol to agree on the destination account and packet condition.
+
+1. Sender constructs an ILP Prepare packet with the receiver's destination account, the agreed-upon condition, and an amount and expiry of their choice. The sender may include additional data, the formatting of which is determined by the higher-level protocol.
+2. Sender sends the ILP Prepare packet to a connector over some communication channel, such as a Websocket. 
+3. Connector gets the packet and determines who the sender was based on the communication channel it was received through. The connector checks whether the sender has sufficient balance or credit to send the amount specified in the packet. If so, the connector reduces the sender's available balance by the value of the packet. If not, the connector returns an ILP Reject packet.
+4. Connector uses the destination ILP address from the packet and its local routing tables to determine which next hop to forward the packet to.
+5. Connector modifies the packet amount and expiry to apply their exchange rate and move the expiry timestamp earlier.
+6. Connector forwards the packet to the next hop, which may be another connector. All subsequent connectors go through steps 3-6 (treating the previous connector as the sender) until the packet reaches the receiver.
+7. Receiver checks the ILP Prepare packet, according to whatever is stipulated by the higher-level protocol (such as checking whether the amount received is above some minimum specified by the sender). 
+8. Receiver returns an ILP Fulfill packet, containing the preimage of the condition in the ILP Prepare packet. The receiver may include additional data, the formatting of which is determined by the higher-level protocol.
+9. Connector checks that the receiver has returned the fulfillment before the expiry in the ILP Prepare packet. If so, the connector returns the same ILP Fulfill packet to the previous connector and credits the receiver's account with the amount in the ILP Prepare packet. If the Prepare has expired, the connector returns an ILP Reject packet to the previous connector. Each connector repeats this step (treating the next connector as the receiver whom they will credit) until the Fulfill (or Reject) packet reaches the sender.
+10. Sender checks that the preimage in the ILP Fulfill packet matches the condition in their Prepare packet and may read any data returned by the receiver.
+11. Sender signs an update to their payment channel with the first connector to cover the value of the ILP Prepare (or they may settle less frequently using slower on-ledger transfers or physical deliveries of cash or other goods) and communicates it to the connector.
```

Comment:
Not sure if 11 and 12 should actually go in here, because technically they're not a part of the packet flow. Maybe an appendix of some kind to this list? I guess for payment channels it makes sense that you might do it every packet, but that's not mandated by ILP

Reply:
I think it's useful to have in there to give some explanation of how the balance is expected to be settled.

> I guess for payment channels it makes sense that you might do it every packet, but that's not mandated by ILP

Is that not clear from "(or they may settle less frequently using slower on-ledger transfers or physical deliveries of cash or other goods)"?

Reply:
I agree with @sharafian here. We should describe it as part of the overall process but not put it in the flow. This may be a pre-paid account in which case this step happens before step 2.